### PR TITLE
233 e2e and functional test for Busy Indicator

### LINKED ISSUE #233: SOHO-8161: Busy Indicator Functional and E2E Tests
Add some tests for Busy Indicator 

See existing tests for example textarea, dropdown, treemap, menubutton for reference. Also see docs/TESTING.MD for commands. Ask for help on sohoxi-dev if needed.

Functional Tests
- Test events that get triggered
- Test events that are listened to
- Test all API methods, most importantly the public ones first and the rest for coverage. (fx enable/disable)
- Test lifecycle methods (init, teardown, destroy, invoke)
- Test setting and that some states are applied

E2e Tests
- Load each test page in the components/<name>/list and test the main ideas of the page, if unsure of the main idea ask on the channel.
- Keyboard

Busy Indicator  Specific Test Considerations 
- show on body
- customizing text
- on inputs
- non blocking
- on an ajax call
- block ui
- block part area
- block whole page
- contains in font-size 0
- delaying display
- on a tab
- nested
- updating

Other Tests
- Accessibility tests https://github.com/dequelabs/axe-core
- Security Tests
- Specific Test Cases from Cindy
- Protractor Image Comparison Tests

### PR DESCRIPTION
**Pull request by tmcastro**
**Explain the _details_ for making this change. What existing problem does the pull request solve?**
- Added e2e and functional test scripts for Busy Indicator

**Related github/jira issue (required)**:
Closes #233 

**Steps necessary to review your pull request (required)**:
- Run `npm run functional:local` on terminal
- Run `npm run e2e:ci` on terminal
- Busy indicator should now be added

### CODE REVIEW

File: test/components/busyindicator/busyindicator.e2e-spec.js [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    await buttonEl.click();
+
+    await browser.driver.wait(protractor.ExpectedConditions.presenceOf(element(by.className('busy-indicator-container'))), config.waitsFor);
+
+    expect(await element(by.className('busy-indicator-container'))).toBeTruthy();
+  });
+});
+
+describe('Busy Indicator example-custom-loading-text tests', () => {
+    beforeEach(async () => {
```
Comment (EdwardCoyle) [inline test/components/busyindicator/busyindicator.e2e-spec.js]: @tmcastro this is one of the reasons the tests failed - our linter doesn't like the spacing here:  https://travis-ci.com/infor-design/enterprise/jobs/144096079#L514

File: test/components/busyindicator/busyindicator-api.func-spec.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
+import { Button, BusyIndicator } from '../../../src/components/busyindicator/busyindicator';
```
Comment (EdwardCoyle) [inline test/components/busyindicator/busyindicator-api.func-spec.js]: Should remove `Button` from here (not exported by the busy indicator source code)

File: test/components/busyindicator/busyindicator-api.func-spec.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+const busyindicatorHTML = require('../../../app/views/components/busyindicator/example-index.html');
+const svg = require('../../../src/components/icons/svg.html');
+
+let busyindicatorEl;
+let svgEl;
+let busyindicatorObj;
+
+describe('Busy Indicator API', () => {
```
Comment (EdwardCoyle) [inline test/components/busyindicator/busyindicator-api.func-spec.js]: Seems that all four of these new functional tests are failing on a check for `classList` returning null:  https://travis-ci.com/infor-design/enterprise/jobs/144096078#L2293

Tested it locally, and it appears to be the same.  I debugged the tests a bit here and it seems like it's looking for an element on that page with the class `.busy-indicator`, where one doesn't exist (yet).

screen shot 2018-09-06 at 1 50 27 pm

If you check out [that test page](http://master-enterprise.demo.design.infor.com/components/busyindicator/example-index.html), the busy indicator UI-blocking element isn't created when the page loads, but when the submit button's clicked.  You probably need to adjust the tests' `beforeEach()` to account for the button click.

**Comment by tmcconechy** [comment]
Also i would merge master onto your branch as it should fix some tests.

File: test/components/busyindicator/busyindicator.e2e-spec.js [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+requireHelper('rejection');
+
+jasmine.getEnv().addReporter(browserStackErrorReporter);
+
+describe('Busy Indicator example-index tests', () => {
+  beforeEach(async () => {
+    await utils.setPage('/components/busyindicator/example-index');
+  });
+
+  it('Should not have errors', async () => {
```
Comment (tmcconechy) [inline test/components/busyindicator/busyindicator.e2e-spec.js]: Empty function.. You can actually add this test to each describe in the e2e tests.

```
  it('Should not have errors', async () => {
    await utils.checkForErrors();
  });
```

File: test/components/busyindicator/busyindicator.e2e-spec.js [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    await utils.setPage('/components/busyindicator/example-custom-loading-text');
+  });
+
+  it('Should display busy indicator with custom text', async () => {
+    const buttonEl = await element(by.id('submit'));
+    await buttonEl.click();
+
+    const customTextEl = await element(by.className('busy-indicator-container')).element(by.tagName('span'));
+
+    expect(customTextEl.getText()).toEqual('Hang Tough, Skippy...');
```
Comment (tmcconechy) [inline test/components/busyindicator/busyindicator.e2e-spec.js]: Missing await

File: test/components/busyindicator/busyindicator.e2e-spec.js [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+  it('Should display busy indicator inside input', async () => {
+    const inputLoadBtn = await element(by.id('trigger-busy-input'));
+
+    await inputLoadBtn.click();
+
+    const fieldEl = await element.all(by.className('field')).get(0);
+
+    expect(await fieldEl.element(by.className('busy-indicator-container'))).toBeTruthy();
+
+    // const ec = protractor.ExpectedConditions;
```
Comment (tmcconechy) [inline test/components/busyindicator/busyindicator.e2e-spec.js]: Remove if not needed.

File: test/components/busyindicator/busyindicator.e2e-spec.js [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+  it('Should display busy indicator inside dropdown', async () => {
+    const dropdownLoadBtn = await element(by.id('trigger-busy-dropdown'));
+
+    await dropdownLoadBtn.click();
+
+    const fieldEl = await element.all(by.className('field')).get(1);
+
+    expect(await fieldEl.element(by.className('busy-indicator-container'))).toBeTruthy();
+
+    // await browser.driver.wait(protractor.ExpectedConditions.stalenessOf(element(by.className('busy-indicator-container'))), config.waitsFor);
```
Comment (tmcconechy) [inline test/components/busyindicator/busyindicator.e2e-spec.js]: remove if not needed.

File: test/components/busyindicator/busyindicator.e2e-spec.js [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+describe('Busy Indicator example-non-blocking tests', () => {
+  beforeEach(async () => {
+    await utils.setPage('/components/busyindicator/example-non-blocking');
+  });
+
+  it('Should display non-blocking busy indicator', async () => {
+    const startBtn = await element(by.id('busy-start-trigger'));
+    await startBtn.click();
+
+    // Check difference between blocking and non-blocking
```
Comment (tmcconechy) [inline test/components/busyindicator/busyindicator.e2e-spec.js]: Missing check

File: test/components/busyindicator/busyindicator.e2e-spec.js [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+  beforeEach(async () => {
+    await utils.setPage('/components/busyindicator/test-block-entire-ui');
+  });
+
+  it('Should block entire UI', async () => {
+    const buttonEl = await element(by.id('submit'));
+    await buttonEl.click();
+
+    const addressEl = await element(by.id('busy-field-address'));
+    addressEl.sendKeys(protractor.Key.NUMPAD1);
```
Comment (tmcconechy) [inline test/components/busyindicator/busyindicator.e2e-spec.js]: Missing await

File: test/components/busyindicator/busyindicator.e2e-spec.js [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+});
+
+describe('Busy Indicator test-block-specific-area tests', () => {
+  beforeEach(async () => {
+    await utils.setPage('/components/busyindicator/test-block-specific-area');
+  });
+
+  it('Should block specific area', async () => {
+    const startBtn = await element(by.id('busy-start-trigger'));
+    // const completeBtn = await element(by.id('busy-complete-trigger'));
```
Comment (tmcconechy) [inline test/components/busyindicator/busyindicator.e2e-spec.js]: remove if not needed

File: test/components/busyindicator/busyindicator.e2e-spec.js [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+    const specificEl = await element(by.id('standalone-busy'));
+
+    await startBtn.click();
+
+    expect(await specificEl.element(by.className('busy-indicator-container'))).toBeTruthy();
+
+    // await browser.actions().click(completeBtn).perform();
```
Comment (tmcconechy) [inline test/components/busyindicator/busyindicator.e2e-spec.js]: remove

File: test/components/busyindicator/busyindicator.e2e-spec.js [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+describe('Busy Indicator test-contained-in-font-size-0 tests', () => {
+  beforeEach(async () => {
+    await utils.setPage('/components/busyindicator/test-contained-in-font-size-0');
+  });
+
+  it('Should display normal font', async () => {
+    const startBtn = await element(by.id('busy-start-trigger'));
+    await startBtn.click();
+
+    const fontSize = element(by.className('busy-indicator-container')).element(by.tagName('span')).getCssValue('font-size');
```
Comment (tmcconechy) [inline test/components/busyindicator/busyindicator.e2e-spec.js]: missing await

File: test/components/busyindicator/busyindicator.e2e-spec.js [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+    const buttonEl = await element(by.id('submit'));
+
+    await buttonEl.click();
+
+    const customTextEl = await element(by.className('busy-indicator-container')).element(by.tagName('span'));
+
+    const EC = protractor.ExpectedConditions;
+    browser.wait(EC.textToBePresentInElement(customTextEl, 'New Text 1'), 5000);
+
+    expect(customTextEl.getText()).toEqual('New Text 1');
```
Comment (tmcconechy) [inline test/components/busyindicator/busyindicator.e2e-spec.js]: missing await

File: test/components/busyindicator/busyindicator-api.func-spec.js [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+  it('Should destroy busy indicator', () => {
+    const buttonEl = $('#submit');
+    buttonEl.click();
+
+    setTimeout(() => {
+      busyindicatorObj.destroy();
+
+      expect(busyindicatorObj).toBeFalsy();
+    }, 1000);
```
Comment (tmcconechy) [inline test/components/busyindicator/busyindicator-api.func-spec.js]: Is such a long time needed? Maybe , 0 or ,1 ?

File: test/components/busyindicator/busyindicator-api.func-spec.js [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+    busyindicatorEl = null;
+    document.body.insertAdjacentHTML('afterbegin', busyindicatorHTML);
+
+    busyindicatorEl = $('#busy-form');
+
+    busyindicatorObj = busyindicatorEl.data('busyindicator');
+  });
+
+  afterEach(() => {
+    if (busyindicatorObj) {
```
Comment (tmcconechy) [inline test/components/busyindicator/busyindicator-api.func-spec.js]: I would suggest adding some code to remove the busy indicator you added to the DOM.
For example https://github.com/infor-design/enterprise/blob/master/test/components/autocomplete/autocomplete-api.func-spec.js#L36

File: test/components/busyindicator/busyindicator-api.func-spec.js [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+
+  it('Should hide busy indicator when triggerring "complete.busyindicator"', () => {
+    const buttonEl = $('#submit');
+    buttonEl.click();
+
+    setTimeout(() => {
+      busyindicatorEl.trigger('complete.busyindicator');
+
+      expect(document.body.querySelector('.busy-indicator')).toBeFalsy();
+    });
```
Comment (tmcconechy) [inline test/components/busyindicator/busyindicator-api.func-spec.js]: Is this missing a time?

File: test/components/busyindicator/busyindicator-api.func-spec.js [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+  it('Should update text of busy indicator', () => {
+    const buttonEl = $('#submit');
+    buttonEl.click();
+
+    let customText = '';
+
+    setTimeout(() => {
+      customText = 'Custom Text';
+      busyindicatorObj.updated({ text: customText });
+    }, 1000);
```
Comment (tmcconechy) [inline test/components/busyindicator/busyindicator-api.func-spec.js]: Can this be shortened

File: test/components/busyindicator/busyindicator-api.func-spec.js [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+
+    setTimeout(() => {
+      const customTextEl = $('.busy-indicator-container > span');
+
+      expect(customTextEl.text()).toEqual(customText);
+    }, 2000);
```
Comment (tmcconechy) [inline test/components/busyindicator/busyindicator-api.func-spec.js]: And this?

File: test/components/busyindicator/busyindicator-api.func-spec.js [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+  });
+
+  it('Should return correct value for isActive', () => {
+    const buttonEl = $('#submit');
+    buttonEl.click();
+
+    setTimeout(() => {
+      expect(busyindicatorObj.isActive()).toEqual(true);
+    }, 1000);
```
Comment (tmcconechy) [inline test/components/busyindicator/busyindicator-api.func-spec.js]: And this?

File: test/components/busyindicator/busyindicator.e2e-spec.js [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+    await utils.checkForErrors();
+  });
+
+  it('Should display busy indicator with custom text', async () => {
+    const buttonEl = await element(by.id('submit'));
+    await buttonEl.click();
+
+    const customTextEl = await element(by.className('busy-indicator-container')).element(by.tagName('span'));
+
+    expect(customTextEl.getText()).toEqual('Hang Tough, Skippy...');
```
Comment (tmcconechy) [inline test/components/busyindicator/busyindicator.e2e-spec.js]: Missing await

File: test/components/busyindicator/busyindicator.e2e-spec.js [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+  });
+
+  it('Should block entire UI', async () => {
+    const buttonEl = await element(by.id('submit'));
+    await buttonEl.click();
+
+    const addressEl = await element(by.id('busy-field-address'));
+    await addressEl.sendKeys(protractor.Key.NUMPAD1);
+
+    expect(addressEl.getText()).not.toEqual('1');
```
Comment (tmcconechy) [inline test/components/busyindicator/busyindicator.e2e-spec.js]: Missing await

File: test/components/busyindicator/busyindicator.e2e-spec.js [hunk 22]
```diff
@@ -0,0 +1,254 @@
...
+  });
+
+  it('Should not have errors', async () => {
+    await utils.checkForErrors();
+  });
+
+  it('Should block whole page', async () => {
+    const busyindicatorEl = await element(by.css('#maincontent > .busy-indicator-container'));
+
+    expect(busyindicatorEl).toBeTruthy();
```
Comment (tmcconechy) [inline test/components/busyindicator/busyindicator.e2e-spec.js]: Missing await - also you can requery 

```
 expect(await element(by.css('#maincontent > .busy-indicator-container')).toBeTruthy();
```

File: test/components/busyindicator/busyindicator.e2e-spec.js [hunk 23]
```diff
@@ -0,0 +1,254 @@
...
+
+  it('Should not have errors', async () => {
+    await utils.checkForErrors();
+  });
+
+  it('Should display normal font', async () => {
+    const startBtn = await element(by.id('busy-start-trigger'));
+    await startBtn.click();
+
+    const fontSize = element(by.className('busy-indicator-container')).element(by.tagName('span')).getCssValue('font-size');
```
Comment (tmcconechy) [inline test/components/busyindicator/busyindicator.e2e-spec.js]: missing await

File: test/components/busyindicator/busyindicator.e2e-spec.js [hunk 24]
```diff
@@ -0,0 +1,254 @@
...
+    await nestedStartBtn.click();
+    await buttonEl.click();
+
+    const nestedStandaloneEl = await element(by.id('nested-busy-standalone'));
+
+    expect(await nestedStandaloneEl.element(by.className('busy-indicator-container'))).toBeTruthy();
+
+    const nestedBusyIndicatorEl = await element.all(by.css('.busy-indicator-container'));
+
+    expect(nestedBusyIndicatorEl.length).toEqual(2);
```
Comment (tmcconechy) [inline test/components/busyindicator/busyindicator.e2e-spec.js]: missing await? can requery if giving stale messages

**Comment by tmcconechy** [comment]
Also one test is failing https://travis-ci.com/infor-design/enterprise/jobs/146050457#L2252 it must be doing a form submit in the test?

**Pull request merged**